Bit-packed GF2

amirebrahimi opened the pull request:
Initially, I started with an approach that defined bit-packing at `ArrayMeta`, so it could be used in `Array` and `FieldArray`. Then, I would handle implicit bit packing and unpacking as necessary depending on the operation. This soon turned into a game of whack-a-mole where I was needing to override a lot of numpy functions on top of ndarray methods.

This version is much simpler, smaller scope, and has limited functionality: basically arithmetic operations and matrix-vector/matrix-matrix multiplication. I figure let's start with this and then expand scope as needed.

Rather than introduce fields higher up in the class/meta hierarchy I've opted for overriding specific functionality in a custom `GF2BP` class. There are simple conversions for going between `GF2` and `GF2BP` via `.astype`, which implies that you will receive a new array.

The numpy v2 and numba v0.60 bump was necessary, so that I could make use of `np.bitwise_count`. This allows a roughly **4x** speed-up in matrix-matrix multiplication.

I started to work on tests and tried to introduce `GF2BP` under the `FIELDS` fixture construction, but there won't be an easy way to `xfail` the tests I know would fail. Ideally, I'd like to piggyback on the tests in `test_arithmetic.py` and `test_linalg.py`, but only in comparison to `GF2` after unpacking. So, let me know if you have any good ideas for this without having to duplicate tests.

amirebrahimi commented:
Thanks, @mhostetter. I'll work on incorporating this feedback in. One thing I wanted to clarify since it seems to be a misunderstanding. `GF2BP` does the bit-packing for you, so it doesn't require your matrices to already be bit-packed.

mhostetter commented:
> One thing I wanted to clarify since it seems to be a misunderstanding. GF2BP does the bit-packing for you, so it doesn't require your matrices to already be bit-packed. With that information, does that change any of your feedback?

That is my understanding of your current implementation, but also something I'm suggesting we change. To me, having an internal data conversion seems dubious. What is the motivation for doing the conversions internally? In my proposal, if a user wanted to start with an unpacked array, they would `x = galois.GF2(unpacked_array)` and then `x_bp = np.packbits(x)`, which returns a `GF2BP`. If they want to start with a bit-packed array, they would `x_bp = galois.GF2BP(packed_array)`.

My issue with the internal (mandatory) data conversion is that it is not idempotent. For instance, doing `x_bp = galois.GF2BP(x_bp)` repeatedly could change the value of `x_bp`. This isn't true of any of the other Galois field arrays. Here's an example:

```python
x = np.packbits([0, 0, 0, 1])
print(x)
# [16]
x = np.packbits(x)
print(x)
# [128]
x = np.packbits(x)
print(x)
# [128]

mhostetter commented:
Also, I'll try to think about an easy way to plug into the test infrastructure.

amirebrahimi commented:
To your question about whether it's necessary to track the unpacked shape. I think so because otherwise when you unpack you won't be able to supply the `count` property and will wind up with potentially extra padded zeros.

mhostetter commented:
> To your question about whether it's necessary to track the unpacked shape. I think so because otherwise when you unpack you won't be able to supply the count property and will wind up with potentially extra padded zeros.

Yes, I agree with that. I was thinking we could populate a property like `._count` (or whatever name makes sense) when the array is created inside the override to `np.packbits()`.

So the user has `x = galois.GF2.Random((100, 6))` and then calls `x_bp = np.packbits(x)`. Inside the intercepted call to `packbits()` we can `x_bp = np.packbits(x)` and then `x_bp = x_bp.view(galois.GF2BP)` and then set `x_bp._count = -2`.

Later when the user wants to unpack it, calling `y = np.unpackbits(x_bp)`, we can use the internal `._count` property if `count=None`. This way the unpacked array has the correct original shape. So rather than tracking the shape of the unpacked array, we just need to know how many zeros were appended on the last axis when the array was packed.

I can help you implement the interception of `packbits()` and `unpackbits()`, if you'd like. It might be a bit tricky.

amirebrahimi commented:
> Yes, I agree with that. I was thinking we could populate a property like `._count` (or whatever name makes sense) when the array is created inside the override to `np.packbits()`.

I've implemented `packbits` and `unpackbits`. 
 
> So the user has `x = galois.GF2.Random((100, 6))` and then calls `x_bp = np.packbits(x)`. Inside the intercepted call to `packbits()` we can `x_bp = np.packbits(x)` and then `x_bp = x_bp.view(galois.GF2BP)` and then set `x_bp._count = -2`.
> 
> Later when the user wants to unpack it, calling `y = np.unpackbits(x_bp)`, we can use the internal `._count` property if `count=None`. This way the unpacked array has the correct original shape. So rather than tracking the shape of the unpacked array, we just need to know how many zeros were appended on the last axis when the array was packed.

I'm now tracking the axis element count because it seemed more straightforward. Is there a specific reason why you wanted to track the appended zero count?

amirebrahimi commented:
> If we use `np.packbits()`, I don't think it's necessary to keep track of the unpacked shape, do you? Also, in the code snippet below, I don't believe the output unpacked shape equals the first input's unpacked shape. This is because of NumPy's broadcasting rules.
> 
> ```python
>     def __call__(self, ufunc, method, inputs, kwargs, meta):
>         output = super().__call__(ufunc, method, inputs, kwargs, meta)
>         output._unpacked_shape = inputs[0]._unpacked_shape
>         return output
> ```

Coming back to this. I was trying to find the case of broadcasting where the output shape would not be the same as the input shape and couldn't. Numpy will complain about the broadcasting if the shapes aren't compatible both for the unpacked and packed case. 

```python
a = GF2.Random((10, 10))
b = GF2.Random((1, 10))
x = np.packbits(a)
y = np.packbits(b)
print((a + b).shape)
print((x + y).shape)
```

If the shapes aren't broadcastable, then it will throw a `ValueError`:
```python
a = GF2.Random((10, 10))
b = GF2.Random((2, 10))
x = np.packbits(a)
y = np.packbits(b)
print((b + a).shape)
print((y + x).shape)
```

Was your concern with how I'm choosing the first input, which would cause an issue? If so, then I think a max over the inputs `._axis_count` would work.

mhostetter commented:
> Was your concern with how I'm choosing the first input, which would cause an issue if the operands were switched (as in the last example)? If so, then I think a max over the inputs ._axis_count would work.

Yes, that was a concern.

> Coming back to this. I was trying to find a case of broadcasting where the output shape would not be the same as the input shape and couldn't. Numpy will complain about the broadcasting if the shapes aren't compatible both for the unpacked and packed case.

Using ufunc methods, like `outer` or  `accumulate`, can change the broadcasted shape.

```python
import numpy as np

a = np.random.randint(0, 2, 10)
b = np.random.randint(0, 2, 10)
x = np.packbits(a)
y = np.packbits(b)
print((a * b).shape)
print((x * y).shape)
print(np.multiply.outer(a, b).shape)
print(np.multiply.outer(x, y).shape)
# (10,)
# (2,)
# (10, 10)
# (2, 2)
```

Also, here's crazy example of broadcasting over multiple axes.

```python
import numpy as np

a = np.random.randint(0, 2, (1, 2, 3))
b = np.random.randint(0, 2, (2, 2, 1))
x = np.packbits(a)
y = np.packbits(b)
print((a * b).shape)
print((x * y).shape)
print(np.multiply.outer(a, b).shape)
print(np.multiply.outer(x, y).shape)
# (2, 2, 3)
# (1,)
# (1, 2, 3, 2, 2, 1)
# (1, 1)
```

I think the question is, can we track the axis elements (or appended zeros) through all of the broadcasting and manipulation?

amirebrahimi commented:
> I think the question is, can we track the axis elements (or appended zeros) through all of the broadcasting and manipulation?

Yes, that is a good question. I'll explore that soon.

Just wanted to give you an update on where I'm at. I started looking at making a bit-packed version of `inv`, which then had me look at `concatenate`, which has ultimately led me to consider indexing for both getting items and setting items. At first I started with an explicit version (i.e. separate methods that would need to be called), leaving the default `__getitem__` and `__setitem__` intact, but now I've decided to see how things work if I move back towards an implicit unpacking.

Currently, I've got [index update rules](https://github.com/mhostetter/galois/blob/c07ba0f462d5171da3256e3a698953c5deb5b0b5/src/galois/_fields/_gf2.py#L452) to handle indexing without defaulting to unpacking the whole matrix. I think I can simplify these rules if I convert them into a canonical form first. There's a test file I had Chat GPT generate to provide some examples of all the ways one can index.

My thought for going down this path is I don't see much utility to the class if there aren't at least some of the more common numpy functions supported in bit-packed form. For us, matrix inverses are one of those.

amirebrahimi commented:
Finished a bitpacked implementation of `np.linalg.inv`:

```python
# %%
a = GF2.Random((1024, 1024), seed=4)
x = np.packbits(a)
%timeit np.linalg.inv(a)  # ~3s
%timeit np.linalg.inv(x)  # ~734ms
assert np.array_equal(a, np.unpackbits(x))
```
So, there's a 4.5x speedup w/ 1k matrices.

Also, did some tests in general to see if unpacking portions of matrices was worthwhile over unpacking the full matrix (to validate having a custom `__getitem__` and `__setitem__`:

```python
# %%
M, N = 1800, 1000
a = np.random.randint(0, 2, size=(M, N), dtype=np.uint8)
A = np.packbits(a, axis=-1)  # shape: (M, ceil(N/8))

# ~434 us
%timeit np.unpackbits(A, axis=-1, count=N)

# %%
# ~3 us
%timeit np.unpackbits(A[np.random.randint(0, M)], axis=-1, count=N)
```

>100x speedup, so I think it's worthwhile to keep this approach around. I think if one wants to get at the rawdata, then they just perform a `.view(np.ndarray)`

amirebrahimi commented on tests/fields/test_bitpacked.py:
```diff
@@ -55,11 +55,88 @@ def test_galois_array_indexing():
+    # 9. Using np.newaxis (reshaped array assignment)
+    arr = GF([1, 0, 1, 1])
+    arr = np.packbits(arr)
+    reshaped = arr[:, np.newaxis]  # should this be using arr's data (as would be the case without packbits) or a new array?
```
Still not sure what to do in this case.

amirebrahimi commented:
I've cleaned up the indexing routines and have [started to look at broadcasting](https://github.com/mhostetter/galois/pull/583/commits/913b95df05500d9fad72ff184f72498df4c40ff0). I found a bug in my `inv_bitpacked` implementation that I need to fix first though.

amirebrahimi commented:
I've added and expanded upon your broadcasting examples in `test_bitpacked.py` and have them all passing now. I think I've also sorted out the majority of indexing. There are still more tests to write and to decide what additional functionality might be needed, but we're in a good place for a review again @mhostetter.

rafaelha commented on src/galois/_fields/_gf2.py:
There is an issue with how shapes are tracked. In the following snippet incorrect shapes are returned when packing along axis=1

```python
a = GF2.Random((100, 100))
a_ = np.packbits(a, axis=1)
print(a_.shape)
#  (100, 100)

a_ = np.packbits(a, axis=0)
print(a_.shape)
# (13, 100)
```

amirebrahimi replied:
I've fixed this now and added a new test, `test_shape`.

rafaelha commented on requirements-min.txt:
```diff
@@ -1,13 +1,13 @@
+numba==0.60.0; python_version=="3.9"
 numba==0.55.0; python_version=="3.8"
 numba==0.55.0; python_version=="3.7"
 typing_extensions==4.0.0
```
With this version, I get `ImportError: cannot import name 'Optional' from 'typing_extensions'`

amirebrahimi replied:
I've reverted this change.

rafaelha commented on src/galois/_fields/_gf2.py:
```diff
@@ -152,6 +399,342 @@ class factory :func:`~galois.GF`.
 
 
+@export
+class GF2BP(
```
The GF2BP class should have its own `__repr__`.

amirebrahimi replied:
There's a general function in `_fields\_array.py` to handle printing, so I want to rely on that. I have [added](https://github.com/mhostetter/galois/pull/583/commits/88bea3c5a9c01926524df1251bfa18b3721898a6) printing of "bitpacked" to that function though.

rafaelha commented on src/galois/_fields/_gf2.py:
```diff
@@ -152,6 +399,342 @@ class factory :func:`~galois.GF`.
 
 
+@export
+class GF2BP(
```
Are you planning to allow inserting new axes? The following code produces an error
```python
a = GF2.Random((10, 20))
a_ = np.packbits(a)
a_[:, None]
# ValueError: cannot reshape array of size 30 into shape (10,)
```

amirebrahimi replied:
I've fixed this and added a new test.

BarakKatzir commented on requirements-min.txt:
```diff
@@ -1,13 +1,13 @@
 numpy==1.23.2; python_version=="3.11"
 numpy==1.21.3; python_version=="3.10"
-numpy==1.21.0; python_version=="3.9"
+numpy==2.0.0; python_version=="3.9"
```
I have an old code snippet that might be modified for the `numpy < 2.0`. It's not a drop in replacement for `bitwise_count`, but good for this case where we only want the parity of the count and not the count itself.
It should be reasonably performant, but probably not as good as `bitwise_count`.

```python
import numpy.typing as npt
import numpy as np

def bitwise_parity(arr: npt.NDArray[np.uint8]) -> npt.NDArray[np.bool_]:
    """Return the parity of the number of bits set to 1 in the input array per item."""
    output = arr.copy()
    nbits = np.iinfo(arr.dtype).bits
    for log_shift in range(0, int(np.round(np.log2(nbits)))):
        shift = 2**log_shift
        output = np.bitwise_and(
            np.bitwise_xor(output, np.right_shift(output, shift)),
            np.asarray(int((b"0" * (2 * shift - 1) + b"1") * (nbits // (2 * shift)), 2)).astype(
                arr.dtype
            ),
        )
    return output.astype(bool)
```
It's been a while since I looked at this code, but I remember that this is good for any uint and not just uint8.
The idea is to repeatedly do `bitwise_xor` with each element with itself shifted by powers of 2. That way only log2(num bits) of `bitwise_xor`s are needed.

amirebrahimi replied:
I tried this out in b45ad89 and then tested it with the following code:

```python
import numpy as np
from galois import GF2

N = 4096
g = GF2.Random((N, N))
k = GF2.Random((N, N))

x = np.packbits(g)
y = np.packbits(k)

# %%
%timeit z = x @ y
```

Unfortunately, the custom function isn't better than unpacking the dot product result.

```bash
_bitwise_parity_numpy_v1
1min 5s ± 401 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

v2
32.4 s ± 855 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

unpacking after 
54.4 s ± 799 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

amirebrahimi commented on src/galois/_fields/_gf2.py:
```diff
@@ -152,6 +399,342 @@ class factory :func:`~galois.GF`.
+    def __init__(
+        self,
+        x: ElementLike | ArrayLike,
+        axis_element_count: Optional[int] = None,
```
I've started tracking the axis.

amirebrahimi commented on src/galois/_fields/_gf2.py:
```diff
@@ -99,6 +134,218 @@ def __init_subclass__(cls) -> None:
+
+        assert isinstance(a, GF2BP) and isinstance(b, GF2BP)
+
+        # bit-packed matrices have columns packed by default, so unpack the second operand and repack to rows
```
I've updated the [test](https://github.com/mhostetter/galois/pull/583/commits/799b001d42f75fefaf96557fbcc5c60b1041c854#diff-683c41bbdd3c1c541959c314063aba114bd52114ac9cd5270432916fe902441bR185) to pack with different axes, which also tests the above case.

amirebrahimi commented:
@mhostetter, I've come back to this finally and addressed the comments of others and many of the concerns you had. Would it help to have me decline and create a new draft PR?

amirebrahimi commented:
Opening up a new draft PR.
